src/nvim/msgpack_rpc/channel.c

```diff
@@ -436,48 +454,107 @@ static void handle_request(Channel *channel, msgpack_object *request)
 
   if (error.set) {
     // Validation failed, send response with error
-    if (channel_write(channel,
-                      serialize_response(channel->id,
-                                         request_id,
-                                         &error,
-                                         NIL,
-                                         &out_buffer))) {
-      char buf[256];
-      snprintf(buf, sizeof(buf),
-               "Channel %" PRIu64 " sent an invalid message, closed.",
-               channel->id);
-      call_set_error(channel, buf);
-    }
+    request_set_error(&error, channel, request_id);
     return;
   }
 
-  // Retrieve the request handler
-  MsgpackRpcRequestHandler handler;
   msgpack_object *method = msgpack_rpc_method(request);
+  Array args = ARRAY_DICT_INIT;
+  msgpack_rpc_to_array(msgpack_rpc_args(request), &args);
+
+  char *meth_str = NULL;
 
+  // Retrieve the request handler or intercept.
+  bool intercepted = false;          // A provider owns this message.
+  MsgpackRpcRequestHandler handler;  // Unused if intercepted.
   if (method) {
-    handler = msgpack_rpc_get_handler_for(method->via.bin.ptr,
-                                          method->via.bin.size);
+    meth_str = xmemdupz(method->via.bin.ptr, method->via.bin.size);
+    char func[256];
+    snprintf(func, 256, "provider#message#%s", meth_str);
+    if (eval_has_func(func)) {
+      intercepted = true;
+    } else {
+      handler = msgpack_rpc_get_handler_for(method->via.bin.ptr,
+                                            method->via.bin.size);
+    }
   } else {
     handler.fn = msgpack_rpc_handle_missing_method;
     handler.defer = false;
   }
 
-  Array args = ARRAY_DICT_INIT;
-  msgpack_rpc_to_array(msgpack_rpc_args(request), &args);
-  bool defer = (!kv_size(channel->call_stack) && handler.defer);
+  // Always defer intercepted messages since they run vim script.
+  bool defer = intercepted || (!kv_size(channel->call_stack) && handler.defer);
   RequestEvent *event_data = xmalloc(sizeof(RequestEvent));
   event_data->channel = channel;
+  event_data->method = meth_str;
   event_data->handler = handler;
   event_data->args = args;
   event_data->request_id = request_id;
   incref(channel);
   event_push((Event) {
-    .handler = on_request_event,
+    .handler = intercepted ? on_message_event : on_request_event,
     .data = event_data
   }, defer);
 }
 
+static void on_message_event(Event event)
```

Comment:
> should the notification include the contents of the message? Clients could then parse it for "Press ENTER" if they want to respond to what I'm calling PressEnter events.

Messages are localised, parsing their textual content is thus by definition brittle.